### ISSUE TITLE
[macos] Can't grant permissions to desktop app

### ISSUE THREAD

**Issue by chadcatlett** [tkainrad/keycombiner#66]
**Details**
Desktop App Version: 0.5.0
macOS: 10.15.7 19H1519 x86_64

**Describe the bug**
Upload launching of the desktop app it indicates that it needs two sets of permissions. Accessibility and Screen Recording.  Under System Preferences -> Security & Privacy I can see that KeyCombiner has the accessibility permission but I don't see an option to grant it Screen Recording permissions.

**To Reproduce**
1. Launch KeyCombiner
2. See dialog about permissions issues
3. Launch System Preferences -> Security & Privacy and attempt to grant all permissions.

**Screenshots**
Dialog screen:

![dialog](https://user-images.githubusercontent.com/108175/145091791-8ac6d1b0-23ca-43fb-b197-4a3206b3c61b.png)

Accessibility showing the application having permission:

![accessibility permissions](https://user-images.githubusercontent.com/108175/145091936-0331bcc0-0a0c-4128-8547-f48abec24e2f.png)

And showing that the application doesn't show up in Screen Recording:
![screenrecording-part-one](https://user-images.githubusercontent.com/108175/145092038-4d257314-f45e-4b54-b6ca-a19ce9b22f86.png)

![screenrecording-part-two](https://user-images.githubusercontent.com/108175/145093399-0a14f984-ca7c-4fd8-a99b-6b1869462b68.png)

**Additional context**

If you need any additional information feel free to ping me here or via email. I'm going to attempt to see if your electron app has a secret way to enable dev tools and see if I can see any errors around permissions.

Edit: fixed last screenshot

**Comment by tkainrad** [comment]
Hi,

thank you very much for leaving this amazingly detailed problem description! I am really sorry that I didn't respond yet to your previous feedback submission. I just remembered it now on seeing your issue submission that I still wanted to investigate that.

I am quite surprised by this problem. KeyCombiner, like most Electron/Node-based apps that want to detect the active app, uses the [active-win library](https://github.com/sindresorhus/active-win) to get this information. The library uses native system calls and handles permission requests. I didn't see any problems so far.

Unfortunately, I am not sure where to start investigating this. One thing that could bring something to light would be the log file at `~/Library/Logs/keycombiner/main.log`

**Comment by tkainrad** [comment]
It's a long shot, but I have found one reason why KeyCombiner might not show up: https://apple.stackexchange.com/a/401461/436776

If you currently use a Launcher to start KeyCombiner, please try opening it via Finder.

**Comment by chadcatlett** [comment]
I've tried launching it directly from Finder, using open, as well as from a launcher(Alfred) and they all do it. 

I went down the rabbithole of trying to use tccutil to reset permissions related to the `com.keycombiner.app` app bundle and it still didn't fix it. The log is fairly basic, shows access to screen recording is being denied.

```
[2021-12-07 14:57:51.029] [info] Checking for update
[2021-12-07 14:57:51.049] [info] Is trusted accessibility client: true
[2021-12-07 14:57:51.050] [info] Has media access to screen (screen recording): denied
[2021-12-07 14:57:51.802] [info] Generated new staging user ID: c2766551-be06-5193-bb86-aaaab963b0a8
[2021-12-07 14:57:52.421] [info] Update for version 0.5.0 is not available (latest version: 0.5.0, downgrade is disallowed).
[2021-12-07 14:59:07.726] [info] Not logged in. Will redirect to https://keycombiner.com/collecting/dashboard/?loggedin
```

Looking over the unpacked app.asar,  I don't see an obvious way to launch KeyCombiner.app and have it enable Dev Tools, am I correct in that it isn't possible? I'm curious if I could trick it by attempting to invoke getDisplayMedia() and see if macOS shows any kind of alert or not.

I know the version of macOS is getting pretty old by macOS standards, but if you'd like to attempt to troubleshooting it more in-depth I'd be willing to assist, be it this issue or something closer to real-time.

**Comment by tkainrad** [comment]
Thanks for trying to debug this! I was really hoping that it had to do with launching via a launcher. Other things that seemed to have helped people in similar situations with other software are (unsurprisingly) re-installing and rebooting. 

> Looking over the unpacked app.asar, I don't see an obvious way to launch KeyCombiner.app and have it enable Dev Tools, am I correct in that it isn't possible? 

Yes, I think it's not possible to access DevTools in the packaged app.
However, I don't really see how it would help you. You would not have access to the main process which would be needed for invoking the check manually. 

The macOS version shouldn't be a problem. I have had people with versions prior to 10.13. There, it also works but it is required to install [Swift runtime support libraries](https://support.apple.com/kb/DL1998?locale=en_US).

I'd be very happy to troubleshoot this with you via video call or whatever else we can think of. We'd have to do it after Christmas though, I am a bit overloaded until then. Maybe we can even think of something else until then.

**Comment by ddetton** [comment]
Same problem here on OSX 10.15.7 in the Desktop and web browser.  I am just trying to share a collection with a Mac user.  Should he be able to access the shared collection on a Mac and be able to interact with the keyboard layouts that were built on a windows 11 machine?

**Comment by tkainrad** [comment]
Hi @ddetton, that's a different issue, right? I mean not related to the screen recording permission check not showing up?

But you sure have a good point. Honestly, I didn't think too much about this use case - manually creating a collection on Windows and sharing it with a macOS user. If the collection was created by importing from public collections, it would also include macOS keys.

The first thing the macOS user can do is to enable the column for the Windows keys: 
![image](https://user-images.githubusercontent.com/7394822/147213680-b5079efd-6a4c-4961-b262-a8a971a4f0fd.png)

Then, he will at least see the keys you have set in the shared collection.

To also interact with the collection visualizer, they would have to change their user agent to Windows. KeyCombiner automatically chooses the column for the respective system for mapping onto the virtual keyboard. To do that, open the browser's developer tools and choose any Windows _User agent_ under _Network conditions_:

![image](https://user-images.githubusercontent.com/7394822/147227904-603f7ae0-5272-499c-9891-ebbd512bc9a7.png)

The Mac user will then see your shared collection exactly the same as you do. 

I am aware of course that this is far from an ideal solution to your problem. I will think about how to improve cross-platform collection-sharing.

**Comment by ddetton** [comment]
Hey @tkainrad, thanks for the quick response.  Keycombiner did not show up for me for Screen Recording either, so +1 for this same issue.

Would you like me to create another issue for cross platform sharing so that the visualizer works in a similar way for the sharing recipient?  Btw, the use case is related to Obsidian and I did start with the public obsidian collection.

Your keyboard maps would make a great addition to Obsidian as a plug-in.  I found Keycombiner from your post on the Obsidian.md forum.  I have always found it difficult to manage hotkeys in Obsidian because it doesn't have the visualizer to easily see what keys are available.  Nice work on the app btw.

**Comment by ddetton** [comment]
Just tested again in Safari on MacOS with the User Agent set to Google Chrome Windows and the visualizer shows the collection the same way it appears on Windows.  Thanks for the workaround!

**Comment by tkainrad** [comment]
Thanks for reporting back, @ddetton! Yes, please create another issue for the cross-platform collection sharing.

I thought about making the collection visualizer into a VSCode plugin, but VSCode does not actually expose the user's key bindings to extensions. I might look into Obsidian in the future, thanks for the suggestion!

Do you have any additional information on the screen recording permissions not showing up? Are you also on 10.15.7?

**Comment by ddetton** [comment]
Yes, also on 10.15.7.

**Comment by tkainrad** [comment]
Ok, thanks for clarifying. Unfortunately, I still don't have a definitive solution.

Some things that can help:
- Making sure that you launch KeyCombiner via Finder
- Uninstall KeyCombiner, reboot, install again and make sure you don't click away the permission dialog when first launching instant lookup
- Resetting permissions for KeyCombiner via `tccutil reset ScreenCapture com.keycombiner.app`
- Making sure _System Integrity Protection_ is enabled via `csrutil status`

Sources:
- https://apple.stackexchange.com/a/401461/436776
- https://apple.stackexchange.com/a/384231/436776

**Comment by tkainrad** [comment]
The new KeyCombiner Desktop version 0.7.0 comes with updates for all its third-party dependencies. I have hopes that this might fix the missing permission check. 

Could you check again @chadcatlett? Maybe with a re-install or at least resetting the permissions via `tccutil`. Unfortunately, I could never reproduce this so I don't have a way to verify whether it is resolved.

**Comment by chadcatlett** [comment]
I've upgraded to the latest version and it doesn't change. Before doing more with tccutil I decided to poke around some in active-win(dow) and see if it would trigger the prompts directly by running it in a terminal. It triggered macOS to ask for permission for the terminal, the permissions that the error dialog states that KeyCombiner would be wanting.

I ran dtrace to watch for KeyCombiner to launch and I didn't see it or any of it's children call any variant of exec, although I did see it reading in the javascript code for the node module.

One thing I don't recall(sadly I didn't save output) was if the output indicated whether or not the process had disabled tracing, by chance do Electron release builds disable tracing at some point?

I decided to give another round of tccutil/sip enablement, delete the app, reboot, etc. This time I did it a bit different, instead of just resetting the relevant services, I did a tccutil reset all. Boy oh boy was that some fun on reboot. 🤣 

Unfortunately the flurry permission requests and redirections to System Preferences was cause for premature celebration as when I launched KeyCombiner.app from Finder it did trigger the accessibility request redirection, but it did not request screen recording permissions nor did it upon activation of the global hotkey. In the security settings under it looks like this:

![image](https://user-images.githubusercontent.com/108175/150624563-41362479-17ff-4cc7-9ad5-2d510b867195.png)

Photo Booth screenshot:
<img width="712" alt="Screen Shot 2022-01-21 at 22 34 02" src="https://user-images.githubusercontent.com/108175/150624771-47c438a9-53e3-4127-a924-6fc1f4c76ac7.png">

Before I pull the trigger on a reinstall, any ideas?

**Comment by tkainrad** [comment]
Thanks a lot for trying all these things! It's very much appreciated.

> by chance do Electron release builds disable tracing at some point?

Honestly, I am not sure. There is [`contentTracing`](https://www.electronjs.org/docs/latest/api/content-tracing) but I think that has to be enabled programmatically/explicitly.

>  it did trigger the accessibility request redirection, but it did not request screen recording permissions nor did it upon activation of the global hotkey.

So you saw the accessibility request for the first time? I guess that's some kind of progress at least :see_no_evil: 

> Before I pull the trigger on a reinstall, any ideas?

What do you mean by _reinstall_, since you already tried deleting the app and installing again? Not sure I'd want you to try anything more drastic than that. I am already feeling bad about not being able to resolve this for you. I will think about using [another package](https://www.npmjs.com/package/mac-screen-capture-permissions) to have more explicit control over the permission requests. Unfortunately, I don't have constant access to a macOS device currently, so this might take a bit.

**Comment by chadcatlett** [comment]
> Thanks a lot for trying all these things! It's very much appreciated.

You are very welcome! I love the concept of KeyCombiner so I'm willing to put in some effort to help you get it working for me if it doesn't intrude too much into your time.

> Honestly, I am not sure. There is [`contentTracing`](https://www.electronjs.org/docs/latest/api/content-tracing) but I think that has to be enabled programmatically/explicitly.

I was specifically meaning disabling the usage of dtrace for instrumentation of the various processes used by Electron. I  thought there was a way for an application to tell dtrace, or at least on macOS's version of it, to not allow tracing of the pid, but it is actually working I'm just not seeing any execution of the active-win main executable.


> So you saw the accessibility request for the first time? I guess that's some kind of progress at least 🙈

Baby steps. :) 
 
 
> What do you mean by _reinstall_, since you already tried deleting the app and installing again? Not sure I'd want you to try anything more drastic than that. I am already feeling bad about not being able to resolve this for you. I will think about using [another package](https://www.npmjs.com/package/mac-screen-capture-permissions) to have more explicit control over the permission requests. Unfortunately, I don't have constant access to a macOS device currently, so this might take a bit.

I was going to consider actually just erasing my machine's ssd via nvme commands and reinstalling macOS to see if a fresh install does it.

The other idea is to use a VM. I'll try that and see if it works and let you know. It might be a few days.

Would you be willing to add support for using an env var to enable devtools? while it won't fix it, it'd at least let me muck around with javascript from within the processes versus doing evil debugger ideas :) 

Slack does this as well as some other high profile apps, so I guess they think the risk is minimal or that no one has noticed it.

I greatly appreciate your help troubleshooting this. The application looks amazing and useful when using it via a global shortcut.

What is your primary platform? Does wayland work?

**Comment by tkainrad** [comment]
Thanks for this detailed response. I am busy this weekend but I think I actually know a solution now.

I will reply in detail on Monday. I think you won't need to invest anymore of your time to get a working version ;)

**Comment by tkainrad** [comment]
> I love the concept of KeyCombiner so I'm willing to put in some effort to help you get it working for me if it doesn't intrude too much into your time.

I am happy to hear that! Reporting problems is definitely not an intrusion for me. I wouldn't be able to build KeyCombiner for all platforms if not for reports like that. It's hard as it is, especially with different operating system variants.* 

> I was going to consider actually just erasing my machine's ssd via nvme commands and reinstalling macOS to see if a fresh install does it.

Glad you didn't do that as I don't think it will be necessary. I have to confess to some fairly obvious problems with the handling of macOS permissions. KeyCombiner Desktop would check both the accessibility and screen recording permissions on startup. If either of them is not granted, `active-win` is never executed. Checking for accessibility is done via [`systemPreferences.isTrustedAccessibilityClient(true)`](https://www.electronjs.org/docs/latest/api/system-preferences#systempreferencesistrustedaccessibilityclientprompt-macos) which also triggers the prompt if permissions aren't granted. 

Checking for screen recording permissions is done via [`systemPreferences.getMediaAccessStatus("screen")`](https://www.electronjs.org/docs/latest/api/system-preferences#systempreferencesgetmediaaccessstatusmediatype-windows-macos). The documentation doesn't say anything about this line triggering the prompt, so I guess this is the problem. However, I am pretty sure that this does actually prompt on macOS Big Sur+, as I do test every update on macOS. There are also several other macOS users I am in frequent contact with who don't have problems. 

Anyway, none of this should matter as I think I don't even need screen recording permissions :see_no_evil: 
When I initially implemented this active-win wasn't so clear about needing it only to access a window's title. KeyCombiner needs only the application name. I built a macOS version that doesn't check for screen recording permissions. If you have already granted accessibility permissions, it should work right away. Please download from here @chadcatlett: https://keycombiner.com/download/KeyCombiner-0.7.1.dmg

It will be a couple of days before I can test it on macOS myself, therefore it's not yet activated for auto-update. You would have to download and install manually. I am pretty sure that it will work as only a couple of lines changed, but you never know. Sorry for asking you to try it out first. 


---



> What is your primary platform? Does wayland work?

My primary platform is Ubuntu. Unfortunately, Wayland does not work currently as [active-win's Linux implementation](https://github.com/sindresorhus/active-win/blob/main/lib/linux.js) depends on X11 utilities (`xwininfo`, `xprop`). Wayland actually makes it quite hard to implement something like that, but at least on Gnome, it should be possible by now. I am thinking about filing a PR on active-win. 

> Would you be willing to add support for using an env var to enable devtools? while it won't fix it, it'd at least let me muck around with javascript from within the processes versus doing evil debugger ideas :)

I don't have any problems with that. I am even thinking about making the desktop app entirely open-source. I will look into enabling devtools via an environment variable.





---
*I didn't expect to have reproducibility issues on all three platforms. However, recently someone reported problems on Windows 11, and none of my Windows machines is eligible for upgrading to 11 :grimacing:

**Comment by tkainrad** [comment]
@chadcatlett did you get a chance to test 0.7.1 on your system? Sorry for the nagging, just trying to make sure you didn't miss my comment by accident. Feel free to ignore.

**Comment by tkainrad** [comment]
Good news, I finally got to test this myself on macOS. 

Tl,dr: the fix already described in [my previous comment](https://github.com/tkainrad/keycombiner/issues/66#issuecomment-1019558489) is now live in KeyCombiner Desktop 0.7.2 :partying_face:  
Please update as usual. It no longer requires screen recording permissions and hence should fix the problem for you @chadcatlett @ddetton 


---

I actually could reproduce this also on macOS 11. After doing `tccutil reset All com.keycombiner.app` KeyCombiner didn't request screen recording permissions anymore as you described. An earlier version of KeyCombiner Desktop did the permissions check correctly. Apparently, macOS kept the permission request even after countless manual removals of the installation and re-installs of new versions.

I am sorry that I initially doubted that this is a problem with KeyCombiner itself.

**Comment by chadcatlett** [comment]
@tkainrad sorry I've been away from a keyboard for a bit. I've just upgraded and it works great.

thanks!

**Comment by tkainrad** [comment]
I am glad to hear that, thanks for reporting back! 

Hopefully, KeyCombiner can now make up for the time you lost troubleshooting this problem.
